### PR TITLE
More Verbose Execption Handling

### PR DESCRIPTION
**Pull request by bobstrecansky**
So I tried to add more verbose exception handling without breaking current functionality (this feels really hacky).  It seems to me that it may be pertinent to break these out and not raise a connection error for each request (as denoted in : https://github.com/kennethreitz/requests/issues/2876)

In the current implementation, a ConnectionError is raised for each call, as well as a max retries message.  It seems like the max retries message error should either be removed or made as more of a general string response statement.  This general string response is here:
https://github.com/kennethreitz/requests/blob/master/requests/adapters.py#L503

Doing so would require removing that line and adding explicit responses within that max retry block, but this might break current functionality if people are depending on the response from requests to make other decisions.

Looking forward to a discussion on this topic.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .coverage
+.cache/
```
Comment (sigmavirus24) [inline .gitignore]: This is unrelated. Please remove it.

File: requests/adapters.py [hunk 2]
```diff
@@ -493,13 +493,12 @@ def send(self, request, stream=False, timeout=None, verify=True, cert=None, prox
                 # TODO: Remove this in 3.0.0: see #2811
                 if not isinstance(e.reason, NewConnectionError):
                     raise ConnectTimeout(e, request=request)
-
-            if isinstance(e.reason, ResponseError):
+            elif isinstance(e.reason, ResponseError):
```
Comment (sigmavirus24) [inline requests/adapters.py]: Why are these being changed to `elif` s? In the event that `e.reason` is a `ConnectTimeoutError` but not a `NewConnectionError` we'll now be unable to re-raise in the event that it's one of these other exeption types.

File: tests/test_requests.py [hunk 3]
```diff
@@ -518,7 +518,8 @@ def test_basicauth_encodes_byte_strings(self):
         ))
     def test_errors(self, url, exception):
         with pytest.raises(exception):
-            requests.get(url, timeout=1)
+            r = requests.get(url, timeout=1)
+            assert r.exception == exception
```
Comment (sigmavirus24) [inline tests/test_requests.py]: We do not attach exceptions to Response objects. Why are you testing this here?

File: .gitignore (hunk 1 above)
Comment (bobstrecansky) [reply .gitignore]: Was also including it to help with the unit tests storage; will open a separate PR for it.

File: requests/adapters.py (hunk 2 above)
Comment (bobstrecansky) [reply requests/adapters.py]: @sigmavirus24 - was just trying to come up with a way to not always also return the connection max retries.  Do you have another idea on how we could handle that?

File: tests/test_requests.py (hunk 3 above)
Comment (bobstrecansky) [reply tests/test_requests.py]: Sorry if i'm not following normal convention.  Is there a method you'd like me to use to test this?

File: requests/adapters.py (hunk 2 above)
Comment (Lukasa) [reply requests/adapters.py]: How does changing `if` to `elif` achieve that?

Comment (bobstrecansky) [reply requests/adapters.py]: I was (probably incorrectly) trying to catch only one exception for each of these, rather than multiple.

Comment (sigmavirus24) [reply requests/adapters.py]: Once we raise a new exception, @bobstrecansky, we won't raise a new exception. If we do not actually raise a new exception then we need these fall-through conditions to maintains backwards compatibility.

File: tests/test_requests.py (hunk 3 above)
Comment (sigmavirus24) [reply tests/test_requests.py]: Well I was more concerned that you were beginning to attach exceptions to Response objects. You don't need this assertion. `with pytest.raises(exception)` is the assertion in this test.

File: requests/adapters.py (hunk 2 above)
Comment (bobstrecansky) [reply requests/adapters.py]: There have been a couple issues opened where users of requests have seen that this clustering of multiple error messages (joining the initial error message alongside the max requests retried) can be confusing to the end user, because it isn't abundantly clear with error message is causing the problem.  Can we come up with a proposed solution for that?  That was my initial intent while writing this PR.

File: tests/test_requests.py (hunk 3 above)
Comment (bobstrecansky) [reply tests/test_requests.py]: Sorry for the confusion, i'll remove this case.

File: requests/adapters.py (hunk 2 above)
Comment (sigmavirus24) [reply requests/adapters.py]: @bobstrecansky while I understand that was your intent and that this was an attempt to fix that, your attempt has done little, if anything, to actually fix that. I think you're missing a critical part of the problem here:

Requests lives at the top-level of several layers of libraries. Immediately below Requests is urllib3. *Most* of the exceptions we handle and then *re-wrap* are urllib3 exceptions. Some of those exceptions are simple exceptions raised by urllib3. Others are exceptions raised by *other* libraries that urllib3 relies on and are re-wrapped. For example, sometimes urllib3 may raise an `SSLError` that is wrapping an exception from the standard library's `ssl` module. Othertimes the `SSLError` may be wrapping an exception from `OpenSSL`. It depends on the user's installation. It's also plausible for urllib3 to be wrapping an exception from `httplib` or `http.client` (depending on Python 2 or Python 3). Beyond that each of those libraries may be wrapping some other library's exception.

The confusing part here stems from two things:

1. Errors wrapped in errors wrapped in errors wrapped in ....
2. Overly generic exceptions being raised in Requests or urllib3 that while technically accurate do not immediately portray the whole story.

**Comment by bobstrecansky** [comment]
Closing this in favor of parsing the execution in the client code.
